### PR TITLE
implement support for `?Sync` callbacks

### LINKED ISSUE #263: `ClientBuilder::on(_any)` should be able to take `!Sync` functions too
There's no reason to keep callbacks under `std::sync::RwLock` if it's `FnMut`, so exclusive access is needed anyways. `std::sync::Mutex` should be used instead as it implements `Sync` regardles if the wrapped value implements `Sync` or not.

### PR DESCRIPTION
**Pull request by ZetaNumbers**
Resolves #263

### CODE REVIEW

File: socketio/src/client/builder.rs [hunk 1]
```diff
@@ -177,11 +177,11 @@ impl ClientBuilder {
     #[allow(unused_mut)]
```
Comment (1c3t3a) [inline socketio/src/client/builder.rs:177]: We can remove this now I assume?

File: socketio/src/client/builder.rs [hunk 2]
```diff
@@ -149,11 +149,11 @@ impl ClientBuilder {
...
```
Comment (1c3t3a) [inline socketio/src/client/builder.rs:149]: Same as below, mutex lock should take a `mut self`.

**Review by 1c3t3a — APPROVED** [review]
LGTM % comments. Thanks for the patch :)

File: socketio/src/client/builder.rs (hunk 2 above)
Comment (ZetaNumbers) [reply socketio/src/client/builder.rs]: not sure what you meant but mut is indeed absolutelly useless, so i removed it

Comment (ZetaNumbers) [reply socketio/src/client/builder.rs]: Perhaps arc and mutex (previously rwlock) are created pretty early. Maybe we could create arc and mutex inside on connect method instead, but builder whould work with pure `Option<Callback>`s?

Comment (1c3t3a) [reply socketio/src/client/builder.rs]: Mh... does this make a difference?

Comment (ZetaNumbers) [reply socketio/src/client/builder.rs]: Codegen, but nothing else i think.

Comment (1c3t3a) [reply socketio/src/client/builder.rs]: Then I would rather leave it as is.

**Comment by 1c3t3a** [comment]
I am sorry, I missunderstood the mut thing here, could you drop commit 5800990? Then we just merge the first one.

**Comment by ZetaNumbers** [comment]
Why drop https://github.com/1c3t3a/rust-socketio/commit/5800990eb9aa31e97f1bcbdcab31da14eb933552?

**Comment by 1c3t3a** [comment]
> Why drop [5800990](https://github.com/1c3t3a/rust-socketio/commit/5800990eb9aa31e97f1bcbdcab31da14eb933552)?

Currently the builder method doesn't need the mut, but it would be possible (and tbh also expected) that the builder needs mut here. If we were not putting mut here, adding it would be a breaking change, so we might as well just have it there for the case that we ever need mut in a future version of the interface. I didn't realize that acquiring the lock and putting something there doesn't need mut as it uses interior mutability. So I thought we could just drop the `allow(unused_mut)` comment which was wrong.

**Comment by ZetaNumbers** [comment]
> If we were not putting mut here, adding it would be a breaking change

It's nothing like that. This mut isn't even a part of the signature, but the part of the definition. You can mutate immutable value by moving it out of the original immutable variable to the mutable one and mutate it however you want.

**Comment by 1c3t3a** [comment]
Mhh fair point because we don't take `&self` here but a moved value. Then it make sense to remove it actually.

**Pull request merged**